src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/FlatArrayIntervalTree`1.cs

```diff
@@ -88,116 +89,121 @@ public static ImmutableIntervalTree<T> CreateFromSorted<TIntrospector>(in TIntro
+            // 
+            //                g
+            //          _____/ \_____
+            //          d           i
+            //       __/ \__       / \
+            //       b     f      h   j
+            //      / \   /
+            //      a c   e 
+            // 
+            // 13 elements:
+            // h, d, l, b, f, j, m, a, c, e, g, i, k
+            // 
+            //                h
+            //          _____/ \_____
+            //          d           l
+            //       __/ \__       / \
+            //       b     f      j   m
+            //      / \   / \    / \
+            //      a c   e g    i k
+            //
+            // The difference in these cases is the 'd' subtree.  We either have:
+            //
+            // 1. enough elements to fill it and start filling its right sibling ('l').  This is the case in the 13
+            //    element tree.
+            //
+            // 2. not enough elements to start filling the right sibling ('i').  This is the case in the 10 element
+            //    tree.
+            //
+            // In both cases, one of the two children of the root will be a perfect tree (the right child in the 10
+            // element case, and the left element in the 13 element case). From the initial discussion, we know that
+            // perfect trees are trivial to handle when we recurse into them.  For the other side, we will get a tree
+            // that we can then recursively apply the more complex logic to.  And the process repeats downwards.
+
+            // The height of the perfect portion of the tree (the rows that are completely full from left to right).
+            // This is '3' in both of the examples above.
+            var perfectPortionHeight = SegmentedArraySortUtils.Log2((uint)subtreeNodeCount + 1);
+
+            // Then number of nodes in the perfect section.  For both trees above this is 7.
+            var perfectSectionNodeCount = (1 << perfectPortionHeight) - 1;
 
-            var midPoint = (startInclusive + endExclusive) / 2;
-            destination[destinationIndex] = new Node(source[midPoint], MaxEndNodeIndex: destinationIndex);
+            // If the entire subtree we're looking at is perfect or not.  It's perfect if every layer is full.
+            // In the above example, both trees are not perfect.
+            var isPerfect = perfectSectionNodeCount == subtreeNodeCount;
 
-            BuildCompleteTreeRecursive(source, destination, startInclusive, midPoint, GetLeftChildIndex(destinationIndex));
-            BuildCompleteTreeRecursive(source, destination, midPoint + 1, endExclusive, GetRightChildIndex(destinationIndex));
+            // The total tree height.  If we're perfect, it's the height of the perfect portion.  Otherwise
+            // it's one higher (to fit the remaining incomplete row).
+            var treeHeight = isPerfect ? perfectPortionHeight : perfectPortionHeight + 1;
```

Comment:
>isPerfect [](http://example.com/codeflow?start=29&length=9)

would it make things a bit simpler later in the code to just return the value if perfect at this point?

Reply:
will do in followup.